custom_metrics/well-known.js

```diff
@@ -0,0 +1,104 @@
+//[well-known]
+// Uncomment the previous line for testing on webpagetest.org
+
+// README! Instructions for adding a new custom metric for the Web Almanac.
+//
+// 1. Refer for instructions for adding a custom metric in almanamc.js.
+// 2. This file has a special case where a custom metric uses 'fetch' and in that case we need to return a promise that resolves to JSON
+// 3. Test your change by following the instructions at https://github.com/HTTPArchive/almanac.httparchive.org/issues/33#issuecomment-502288773.
+// 4. Submit a PR to update this file.
+
+function fetchWithTimeout(url) {
+  var controller = new AbortController();
+  setTimeout(() => {controller.abort()}, 5000);
+  return fetch(url, {signal: controller.signal});
+}
+
+function parseResponse(url, parser) {
+  return fetchWithTimeout(url).then(request => {
+    let resultObj = {};
+    if(!request.redirected && request.status === 200) {
+      resultObj['found'] = true;
+      if(parser) {
+        let promise = parser(request);
+        if (promise) {
+          return promise.then(data => {
+            resultObj['data'] = data;
+            return [url, resultObj];
+          });
+        } else {
+          resultObj['error'] = 'parser did not return a promise';
+          return [url, resultObj];
+        }
+      } else {
+        return [url, resultObj];
+      }
+    } else {
+      resultObj['found'] = false;
+      return [url, resultObj];
+    }
+  });
+}
+
+return Promise.all([
+  // ecommerce
+  parseResponse('/.well-known/assetlinks.json'),
+  parseResponse('/.well-known/apple-app-site-association'),
+  // privacy
+  parseResponse('/.well-known/gpc.json', r => {
+    return r.text().then(text => {
+      let data = {
+        'gpc': null
+      };
+      let gpc_data = JSON.parse(text);
+      if (typeof gpc_data.gpc == 'boolean') {
+        data.gpc = gpc_data.gpc;
+      }
+      return data;
+    });
+  }),
+  // security
+  parseResponse('/robots.txt', r => {
+    return r.text().then(text => {
+      let data = {'disallows': {}};
+      let currUserAgent = null;
+      for(let line of text.split('\n')) {
+        if (line.startsWith('User-agent: ')) {
+          currUserAgent = line.substring(12);
+          if (data['disallows'][currUserAgent] === undefined) {
+            data['disallows'][currUserAgent] = [];
+          }
+        } else if (line.startsWith('Disallow: ')) {
+          data['disallows'][currUserAgent].push(line.substring(10));
```

Comment:
Is this logging too much info? Curious what you plan to do with it. If you're only interested in high-level stats like whether disallows are used at all and for which UAs, or even how many paths are disallowed, is there a way to summarize those stats in the results instead of outputting every path?

Reply:
We would be interested to check how (and if) the login endpoints are disallowed. A regex to check some keywords like "login", "sign-in", and only capture those paths might be ok. @tomvangoethem @nrllh thoughts?

Reply:
Just did a quick check on a dozen of sites. I think we will cover most login pages if we filter the URL on `login`, `log-in`, `signin`, and `sign-in`.

Just let me know if you'd like to have the code updated this way and I'll push the change @SaptakS @tomvangoethem @nrllh.

PS: I also came across a false positive: https://github.com/account-login refers to an account, but is included in `robots.txt` weirdly enough. 🤔

Reply:
Oh, I completely forgot a discussion I had with Tom. We should also filter on `admin`, might be interesting as well.

Reply:
Maybe also `auth`, `sso` and `account`?
Were you planning on matching 1) the whole path, 2) any path segment, or 3) any substring of the path?

Reply:
> Maybe also `auth`, `sso` and `account`?

Could also be interesting!

> Were you planning on matching 1) the whole path, 2) any path segment, or 3) any substring of the path?

I was thinking any substring of the URL (not only path). There's also signin.aws.amazon.com for instance.

Reply:
> > Were you planning on matching 1) the whole path, 2) any path segment, or 3) any substring of the path?
> 
> I was thinking any substring of the URL (not only path). There's also signin.aws.amazon.com for instance.

That implies expanding every path in `robots.txt` to its full path and matching on that? What I'm getting at: [subdomains appear to require their own `robots.txt`](https://developers.google.com/search/docs/advanced/robots/robots_txt). That would mean that, to determine that `signin.aws.amazon.com` cannot be crawled because of `robots.txt`, you'd have to find a wildcard entry `*` in `signin.aws.amazon.com/robots.txt`. That goes beyond just parsing the contents of `robots.txt` and matching on that. (And technically, _any_ entry of `robots.txt` for a domain which contains a keyword will result in a match.)

Reply:
Yeah. Subdomains might be a little bit trickier to get, I agree. But at the least we should do any substring of the path.

Reply:
Hmm, good point... I wasn't aware that subdomains need their own `robots.txt`.

Still, I think there is no need for fetching `robots.txt` from subdomains that contain certain keywords since the idea is for an attacker to probe the local `robots.txt` to collect potentially sensitive endpoints. So just matching any substring of the path will be sufficient for the sec chapter.

Reply:
Indeed, the 'sensitive' subdomains will need to be discovered through other means anyway (subdomain enumeration etc.), so that is a distinct security issue.

Reply:
I'm not sure I understand the security sensitiveness/impact of adding the login/authentication path to `robots.txt`, especially if we're only checking for paths containing known strings. Unless I'm missing something, the only security issue I see is if someone would hide an admin panel behind a random path and put that path in the `robots.txt`.

Reply:
As far as I understand, `robots.txt` could be abused from an automated attacker's perspective. While attackers can probe server responses for variations of `/login`, `/admin`, etc., this would be inefficient. Furthermore, websites could use more unique endpoints for these purposes (that perhaps still contain the `login`/`admin`/... substring).

So in that sense, matching `robots.txt` for these substrings could aid an automated attack to find potentially sensitive endpoints more efficiently, and could identify more unique variations.

That's just my interpretation though, maybe this metric was proposed with another vulnerability in mind.